File: py_version/Q9.py
import re


def main(data):
    vent_map = {}
    count = 0

    for i in range(0, len(data), 2):
        start = (int(data[i].split(',')[0]), int(data[i].split(',')[1]))
        end = (int(data[i+1].split(',')[0]), int(data[i+1].split(',')[1]))

        # |
        if start[0] == end[0]:
            if start[1] > end[1]:
                for i in range(int(end[1]), int(start[1])+1, 1):
                    pos = (int(start[0]), i)
                    if pos in vent_map:
                        vent_map[pos] += 1
                    else:
                        vent_map[pos] = 1
            else:
                for i in range(int(start[1]), int(end[1])+1, 1):
                    pos = (int(start[0]), i)
                    if pos in vent_map:
                        vent_map[pos] += 1
                    else:
                        vent_map[pos] = 1
        # -
        elif start[1] == end[1]:
            if start[0] > end[0]:
                for i in range(int(end[0]), int(start[0])+1, 1):
                    pos = (i, int(start[1]))
                    if pos in vent_map:
                        vent_map[pos] += 1
                    else:
                        vent_map[pos] = 1
            else:
                for i in range(int(start[0]), int(end[0])+1, 1):
                    pos = (i, int(start[1]))
                    if pos in vent_map:
                        vent_map[pos] += 1
                    else:
                        vent_map[pos] = 1
        else:
            pass

    for ele in vent_map.values():
        if ele >= 2:
            count += 1

    return count


if __name__ == '__main__':
    with open('q9input.in') as f:
        data = f.read()
        data = re.split(r' -> |\n', data)

    print(main(data))

File: go_version/Q9.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"regexp"
	"strconv"
	"strings"
	// "reflect"
)

func read_input(fname string) ([]string, error) {
	datas, err := ioutil.ReadFile(fname)
	if err != nil {
		return nil, err
	}

	var data []string
	point := regexp.MustCompile(" -> |\n").Split(string(datas), -1)
	for _, item := range point {
		data = append(data, strings.TrimSpace(item))
	}

	return data, nil
}

func count_overlap_point(data []string) int {
	type Pair struct {
		x, y int
	}

	vent_map := map[Pair]int{}
	var count int = 0

	for i := 0; i < len(data); i += 2 {
		start_x, _ := strconv.Atoi(strings.Split(data[i], ",")[0])
		start_y, _ := strconv.Atoi(strings.Split(data[i], ",")[1])
		end_x, _ := strconv.Atoi(strings.Split(data[i+1], ",")[0])
		end_y, _ := strconv.Atoi(strings.Split(data[i+1], ",")[1])

		start := Pair{start_x, start_y}
		end := Pair{end_x, end_y}

		// |
		if start.x == end.x {
			if start.y > end.y {
				for i := end.y; i < start.y+1; i += 1 {
					pos := Pair{start.x, i}
					if _, ok := vent_map[pos]; ok {
						vent_map[pos] += 1
					} else {
						vent_map[pos] = 1
					}
				}
			} else {
				for i := start.y; i < end.y+1; i += 1 {
					pos := Pair{start.x, i}
					if _, ok := vent_map[pos]; ok {
						vent_map[pos] += 1
					} else {
						vent_map[pos] = 1
					}
				}
			}
			// -
		} else if start.y == end.y {
			if start.x > end.x {
				for i := end.x; i < start.x+1; i += 1 {
					pos := Pair{i, start.y}
					if _, ok := vent_map[pos]; ok {
						vent_map[pos] += 1
					} else {
						vent_map[pos] = 1
					}
				}
			} else {
				for i := start.x; i < end.x+1; i += 1 {
					pos := Pair{i, start.y}
					if _, ok := vent_map[pos]; ok {
						vent_map[pos] += 1
					} else {
						vent_map[pos] = 1
					}
				}
			}
		} else {
		}
	}

	for _, value := range vent_map {
		if value >= 2 {
			count += 1
		}
	}

	return count
}

func main() {
	data, err := read_input("../input/q9input.in")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(count_overlap_point(data))
}
